shipwright/source_control.py

```diff
@@ -52,6 +55,40 @@ def _hexsha(ref):
         return ref.hexsha[:12]
 
 
+def _hash_blob(blob):
+    if blob.hexsha != blob.NULL_HEX_SHA:
+        return blob.hexsha
+    else:
+        return blob.repo.git.hash_object(blob.abspath)
+
+
+def _dirty_suffix(repo, base_paths=['.']):
+    def abspath(path):
+        return os.path.abspath(os.path.join(repo.working_dir, path))
+
+    def hash_blobs(blobs):
+        return [(b.path, _hash_blob(b)) for b in blobs if b]
+
+    def in_paths(path):
+        return any(abspath(path).startswith(abspath(base_path) + os.sep)
+                   for base_path in base_paths)
+
+    diff = repo.head.commit.diff(None)
+    a_hashes = hash_blobs(map(lambda d: d.a_blob, diff))
```

Comment:
lambdas make guido cry.

a_hashes = hash_blobs(d.a_blob for d in diff)
b_hashes = hash_blobs(d.b_blob for d in diff)